### PR TITLE
Sound preference proposal, closes #83

### LINKED ISSUE #83: Sound enable/disable preference
Hi, the same way we currently have nsfw preference, could there be an sound preference?

Simple way is to reflect the preference in `muted` attribute presence
Advanced would be to also transfer the volume level from one video tag to another (javascript accessible property `videotag.volume` with values from 0.0 to 1.0 meaning 0% to 100% volume)

If you agree with proposal, i'd be willing to create PR for your review

### PR DESCRIPTION
**Pull request by smarek**

### CODE REVIEW

**Comment by ubershmekel** [comment]
Thank you fro doing this work! Two requests if you have time:

* Seems to get v.reddit sounds to work we'd need to handle the `hls_url` instead of the fallback url. That's by looking at https://www.reddit.com/r/WhatsWrongWithYourDog/comments/bv7x95/so_our_dog_gets_a_bit_excited_when_asked_about/.json and https://www.reddit.com/r/WhatsWrongWithYourDog/comments/bv7x95/so_our_dog_gets_a_bit_excited_when_asked_about/
* The buttons in the UI are getting too numerous and unwieldy. Something needs to be collapsed there perhaps. I also think I misplaced the `comments` and `image` links. All the checkboxes should be in one row. I don't expect you to to do that in this PR. Just saying I might have to do it if you fix the v.reddit issue.

**Comment by smarek** [comment]
@ubershmekel cool

1) fallback_url links directly the audio, hls_url links m3u8 playlist, which is usually not very compatible with browsers, and I think we would have to integrate additional JS library to handle such playlists.
Do you insist on replacing fallback_url with hls_url ?

This is what I see in `media.reddit_video.` block of JSON url you've linked, maybe the contents changes over time, and you've seen something very different?
```json
"media": {
"reddit_video": {
"fallback_url": "https://v.redd.it/3fqsu5bczj131/DASH_720?source=fallback",
"height": 720,
"width": 1280,
"scrubber_media_url": "https://v.redd.it/3fqsu5bczj131/DASH_96",
"dash_url": "https://v.redd.it/3fqsu5bczj131/DASHPlaylist.mpd",
"duration": 61,
"hls_url": "https://v.redd.it/3fqsu5bczj131/HLSPlaylist.m3u8",
"is_gif": false,
"transcoding_status": "completed"
}
},
```

2) yes, the layout is not very nice, and i've got no idea what to do with it, however that is something I can fix easily, once we agree on what design we want to have

I propose this controls layout:
```
subreddit link | nswf | sound | fullscreen
hotkeys | src | blog | comments | image
auto next form seconds
pagination
```

what do you think?

**Comment by ubershmekel** [comment]
1. I'm not sure I understand your proposal. `https://v.redd.it/3fqsu5bczj131/DASH_720?source=fallback` has no audio as far as I can tell. Do you hear audio at that url?

2. For layout, I'm thinking of separating the types of UI. First row would be links, the second would be the checkboxes.

```
redditp-info | subreddit link | comments | image
nsfw | sound | auto next | duration | fullscreen
pagination
```

Then `redditp-info` would link to the github readme. This would remove ```hotkeys | src | blog``` which would be explained and linked at the github readme.

What do you think of that layout?

File: js/script.js [hunk 1]
```diff
@@ -334,6 +354,7 @@ $(function () {
             // Need to redesign this redditp thing.
             pic.type = imageTypes.gifv;
             pic.url = pic.data.media.reddit_video.fallback_url;
+            pic.sound = pic.url.substring(0,pic.url.lastIndexOf('/'))+"/audio";
```
Comment (smarek) [inline js/script.js]: Audio url

**Comment by smarek** [comment]
1. sorry, audio is available at `fallback_url/audio` see the script.js line 357
No audio is on the media links, from the dog link you posted, if you look in network trace, you can see reddit is loading DASH video of quality and audio separately
2019-06-08-092206_500x213_scrot

2. I like it, my proposal tried to keep some things where they were, but if you're cool with complete reorg, i like it :)

**Comment by ubershmekel** [comment]
This works on v.redd.it urls like https://www.reddit.com/r/WhatsWrongWithYourDog/comments/c3x7cc/doggo_thinks_that_he_is_one_of_them/.json

Does not work on gfycat urls like https://www.reddit.com/r/nextfuckinglevel/comments/c3v7b3/looks_so_effortless_but_he_must_have_trained_very/.json

**Comment by ubershmekel** [comment]
I re-merged and tested this. This works for a file that has video+audio in it, but not for separate video and audio files like reddit videos. Seems only imgur and gfycat will work with this.

I think to make it fully work we'd need to implement something like this JS multiplexing. Which is what I assume reddit video has done: https://medium.com/canal-tech/how-video-streaming-works-on-the-web-an-introduction-7919739f7e1

I'm not sure if we should merge it in when reddit videos don't work.

**Comment by smarek** [comment]
Could you maybe create a branch on which we could propose PRs to fix remaining issues with sound, from mentioned "re-merged" branch you already have? Or if you have staging environment somewhere for others to test the sound with (if not, it can be disabled by default and listed under experimental options)

Also the "multiplexing" is imho not needed, as you can simply provide different URL that contains MP4 container (audio+video) , or provide separate <video and <audio html tags and sync them using https://stackoverflow.com/a/32323192/492624 or similar techniques

**Comment by ubershmekel** [comment]
https://github.com/ubershmekel/redditp/commits/smarek-sound2

**Comment by smarek** [comment]
@ubershmekel before i start working on that, would you be cool with using [dash.js](https://github.com/Dash-Industry-Forum/dash.js) for playing `(json)reddit_video.dash_url` instead of using plain html5 elements?

**Comment by ubershmekel** [comment]
I see the minified file is 577 KB (non-gzipped size). I couldn't find a browser compatibility matrix. Can you explain the cost-benefit? Seems a bit heavy and potentially fragile.

**Comment by smarek** [comment]
Given raw-json
```json
"reddit_video": {
"fallback_url": "https://v.redd.it/iysd6l20k3941/DASH_480?source=fallback",
"height": 480,
"width": 384,
"scrubber_media_url": "https://v.redd.it/iysd6l20k3941/DASH_96",
"dash_url": "https://v.redd.it/iysd6l20k3941/DASHPlaylist.mpd",
"duration": 19,
"hls_url": "https://v.redd.it/iysd6l20k3941/HLSPlaylist.m3u8",
"is_gif": false,
"transcoding_status": "completed"
}
```

We can either do
1) Plain html+js (whatever source data)
2) dash.js (dash_url)
3) hls.js (hls_url)
4) ???
5) (maybe we can do all of them, and fall-back through when given browser does not support one but supports other?)

I've found conflicted sources, but it seems to me that HLS, being Apple technology, has far more lower support than DASH

  - https://bitmovin.com/mpeg-dash-browser-support-device-compatibility/
  - https://caniuse.com/mpeg-dash
  - https://caniuse.com/http-live-streaming
  - https://caniuse.com/mediasource
  - https://github.com/video-dev/hls.js/#compatibility
  - https://github.com/video-dev/hls.js/blob/master/docs/API.md#first-step-setup-and-support

However reddit itself (on my device) opens with HLS source
```html
<video poster="https://external-preview.redd.it/SZJwCdXt4knnZgskQuE2_C_QDyCXQeDNReuwC-lp5aw.png?width=320&amp;crop=smart&amp;format=pjpg&amp;auto=webp&amp;s=53e709f966bf326c52cb3b64b995cdb2525e2bd8" muted="" preload="auto" class="_1EQJpXY7ExS04odI1YBBlj" src="blob:https://www.reddit.com/b59eb125-a56e-46b4-9d40-9bd09d38b4c4">
<source src="https://v.redd.it/iysd6l20k3941/HLSPlaylist.m3u8" type="application/vnd.apple.mpegURL">
</video>
```

I'd be really grateful if someone could investigate the protocol compatibility between HLS/DASH

Relevant tickets
  - https://github.com/Dash-Industry-Forum/dash.js/issues/2055

**Comment by smarek** [comment]
@ubershmekel I'm quite sorry, but I've solved this ticket and related #104 ticket, but the PR is quite a lot changes, because I was not able to develop efficiently, with so much code issues.

Changes i've included along (and the PR is currently mergable without conflicts against master)
- ignoring idea (idea-community/webstorm/phpstorm) dev files
- code formatting, terminating statements with `;`, strict comparison in if statements
- **ability to handle single-post URL** (useful for debugging)
- **fixed distinguish for json/jsonp requests**
- **disable cloud reporting on non-production environment**

I've left the commits un-squashed, but after review, if accepted, i can (or you can) squash some of them together

Also, since I've updated this PR and did not create new one against branch **smarek-sound2**, said branch can be deleted currently

**Comment by smarek** [comment]
Cool, deploy preview works on my end correctly in all test-cases, check the links below:

https://deploy-preview-84--redditp.netlify.com/
https://deploy-preview-84--redditp.netlify.com/r/all
https://deploy-preview-84--redditp.netlify.com/r/WhatsWrongWithYourDog/comments/c3x7cc/doggo_thinks_that_he_is_one_of_them
https://deploy-preview-84--redditp.netlify.com/r/nextfuckinglevel/comments/c3v7b3/looks_so_effortless_but_he_must_have_trained_very
https://deploy-preview-84--redditp.netlify.com/r/random
https://deploy-preview-84--redditp.netlify.com/r/randnsfw
https://deploy-preview-84--redditp.netlify.com/

File: js/script.js [hunk 2]
```diff
@@ -435,7 +464,8 @@ $(function () {
                 $('#titleDiv .collapser').click();
                 break;
             case A_KEY:
-                $("#autoNextSlide").prop("checked", !$("#autoNextSlide").is(':checked'));
+                var $ans = $("#autoNextSlide");
+                $ans.prop("checked", !$ans.is(':checked'));
```
Comment (smarek) [inline js/script.js]: this is lint issue duplicate jquery selector

File: js/script.js [hunk 3]
```diff
@@ -89,14 +90,15 @@ $(function () {
...
+    var getNextSlideIndex = function (currentIndex) {
+        if (!rp.settings.nsfw) {
             // Skip any nsfw if you should
-            for(var i = currentIndex + 1; i < rp.photos.length; i++) {
+            for (var i = currentIndex + 1; i < rp.photos.length; i++) {
                 if (!rp.photos[i].over18) {
                     return i;
                 }
             }
+            return 0;
```
Comment (smarek) [inline js/script.js]: here not handling situation where none of available pictures is non-nsfw caused ajax looping, but it's obscure usecase

File: index.html [hunk 4]
```diff
@@ -41,7 +41,12 @@
                 return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
             });
         }
+
         window.addEventListener('DOMContentLoaded', function () {
+            if (window.location.hostname !== 'redditp.com') {
```
Comment (smarek) [inline index.html]: this exclusion should work correctly in all browsers, i suppose there is no need to send logs from dev or deploy preview environments

File: js/script.js [hunk 5]
```diff
@@ -194,19 +196,35 @@ $(function () {
...
     };
 
+    var updateSound = function () {
+        rp.settings.sound = $('#sound').is(':checked');
+        setCookie(cookieNames.soundCookie, rp.settings.sound);
+        var videoTags = document.getElementsByTagName('video');
+        if (videoTags.length === 1) {
+            videoTags[0].muted = !rp.settings.sound;
+        }
+        var audioTags = document.getElementsByTagName('audio');
```
Comment (smarek) [inline js/script.js]: since for v.redd.it audio tag is separate from video tag, this needs to be checked as well

File: js/script.js [hunk 6]
```diff
@@ -342,14 +370,15 @@ $(function () {
                 // some crossposts don't have a pic.data.media obj?
                 return;
             }
+            pic.sound = pic.url.substring(0, pic.url.lastIndexOf('/')) + "/audio";
```
Comment (smarek) [inline js/script.js]: this generates audio link for v.redd.it links only

File: js/script.js [hunk 7]
```diff
@@ -726,12 +770,7 @@ $(function () {
         }
         var extension = url.substring(dotLocation);
 
-        if (rp.settings.goodExtensions.indexOf(extension) >= 0) {
-            return true;
-        } else {
-            //log("skipped bad extension: " + url);
-            return false;
-        }
+        return rp.settings.goodExtensions.indexOf(extension) >= 0;
```
Comment (smarek) [inline js/script.js]: simplify expression lint

File: js/script.js [hunk 8]
```diff
@@ -786,22 +825,22 @@ $(function () {
...
+    var isShuffleOn = function () {
         var query = parseQuery(window.location.search);
-        console.log("querrry", query);
-        if (query.shuffle) {
-            return true;
-        } else {
-            return false;
-        }
-    }
+        return !!query.shuffle;
```
Comment (smarek) [inline js/script.js]: bang-bang, see https://gist.github.com/mledour/d936fa347a8c3d754c03f2ad8cd96c88

File: js/script.js [hunk 9]
```diff
@@ -786,22 +825,22 @@ $(function () {
...
+    };
 
-    var startShow = function() {
+    var startShow = function () {
         startAnimation(0);
-    }
-    
+    };
+
+    var getStackTrace = function () {
```
Comment (smarek) [inline js/script.js]: helper function, unused currently, but very useful in tracing dev issues

File: js/script.js [hunk 10]
```diff
@@ -811,7 +850,8 @@ $(function () {
         rp.session.loadingNextImages = true;
 
         // Note that JSONP requests require `".json?jsonp=?"` here.
-        var jsonUrl = rp.redditBaseUrl + rp.subredditUrl + ".json?jsonp=?" + rp.session.after + "&" + getVars;
+        var jsonUrl = rp.redditBaseUrl + rp.subredditUrl + ".json?" + (rp.session.after ? rp.session.after + "&" : "") + getVars;
```
Comment (smarek) [inline js/script.js]: jsonp query attribute is added based on below useJsonP classification

**Comment by ubershmekel** [comment]
Looks like a lot of work. What's the state of affairs?

**Comment by smarek** [comment]
What do you mean?

File: index.html (hunk 4 above)
Comment (ubershmekel) [reply index.html]: `window.location.hostname === "www.redditp.com"` when I check.

```suggestion
            if (window.location.hostname.indexOf('redditp.com') === -1) {
```

**Comment by smarek** [comment]
If the single proposed change is everything, i think, it's ready to merge

**Comment by ubershmekel** [comment]
Looking at the code, this is all great stuff. I wish I had a test suite that could run through more scenarios. I hope to finish reviewing tomorrow.

**Comment by smarek** [comment]
Only thing I'm not satisfied with, is the detection of subreddits, that redirect.

There are certainly more than what's hard-coded currently, and I did not find any official list of this kind of subreddits, and detecting whether the call resulted in response from different subreddit (which might not be complete/best solution), is not possible via standard jQuery XHR means (currently being solved here https://github.com/jquery/jquery/pull/4405 )

I propose to solve this in subsequent PR, because ie. `r/random` returns data from randomly picked subreddit, but currently implementation calls next page of items from url such as ` http://www.reddit.com/r/random.json?&after=t3_fap561&` which is not correct obviously

File: js/script.js [hunk 11]
```diff
@@ -892,21 +940,23 @@ $(function () {
...
         };
 
         if (rp.settings.debug)
             console.log('Ajax requesting: ' + jsonUrl);
 
-        // I still haven't been able to catch jsonp 404 events so the timeout
-        // is the current solution sadly.
-        // Note we're still using `jsonp` despite potential issues because
-        // `http://www.redditp.com/r/randnsfw` wasn't working with CORS for some reason.
-        // https://github.com/ubershmekel/redditp/issues/104
```
Comment (ubershmekel) [inline js/script.js]: You removed this comment, without providing an explanation for the `jsonp` toggling or why did you make the following specific exceptions to`useJsonP`. We should pick this up in a future PR.

**Comment by ubershmekel** [comment]
I love that getting a `/comments/` URL works now. I was hoping to get that fixed some day. I'm curious why you need to toggle between jsonp and not for specific URLs. I thought JSONP works for everything. No?

Merging. We'll get the rest sorted later. Thank you so much for all this work @smarek it works beautifully as far as I can tell.

File: js/script.js (hunk 11 above)
Comment (smarek) [reply js/script.js]: Sure, it seemed to me that useJsonP with the code, that decides whether it's true or false, is pretty self-explanatory, but as I said, I expect this to be solved better when we tackle the redirected subreddits.

**Comment by smarek** [comment]
> I love that getting a `/comments/` URL works now. I was hoping to get that fixed some day. I'm curious why you need to toggle between jsonp and not for specific URLs. I thought JSONP works for everything. No?

This is given by response type of the endpoint, when the request is redirected or single-post, the response type does not match application/json, and is blocked by CORB (at least in my chrome dev version 82)

**Pull request merged**